### ISSUE TITLE
SpiNNaker packet mask

### ISSUE THREAD

**Issue by arrenglover** [mcasti-iit/HPU_Core#2]
Additional Feature - 

The data sent and received from the SpiNNaker could have a mask applied to it such that unwanted bits can be removed. Such an operation is *necessary* as the SpiNNaker uses the packet for routing and additional bits in the packet will cause the routing to fail. Additionally any additional bits added to the packet for routing may not be compatible with the codec when packets are pushed into YARP. These masks are not necessarily identical.

It is currently performed in the zynqGrabber and this is a feature that will offload processing to the FPGA

The masks should be controlled by registers. Defaults values
0x00FFFFFF
0x00FFFFFF

**Comment by mcasti-iit** [comment]
Question: the mask should affect every received data, including command keys?
Namely, the default START key comand is 0x80000000 and the default STOP command key is 0x40000000. 
If the mask would affect every received data, with default value 0x00FFFFFF the commands cannot ever be recongnized. 
In other words: the RX mask should be applied to input data coming from SpiNNaker, or to output data provided to Zynq?

**Comment by arrenglover** [comment]
If it is possible the FPGA should recognise the specific commands for "start" and "stop" and if it isn't one of these special commands, it should apply the mask to the data. Is that possible?

**Comment by mcasti-iit** [comment]
Sure, it is possible, and your question gives an answer to my question ;O)
So, I make the filter to be applied to the **output data to ZYNQ processor**; that way, whole bitwidth of data coming from Spinnaker is processed by FPGA (and command recognised despite the mask), and uP (and your application) receives just masked data

**Comment by arrenglover** [comment]
Ideally, I would like 2 masks, one for ZYNQ->spiNNaker (required to make sure it gets routed correctly) and one for spiNNaker->ZYNQ (required as the spinnaker might add bits to route it correctly). These masks are probably the same, but might be different. Is it possible?

**Comment by mcasti-iit** [comment]
I've implemented two different masks, one for TX  and one for RX, settable with two different registers

**Comment by mcasti-iit** [comment]
Go to page
https://github.com/event-driven-robotics/zynq-spinnakerlink/releases/tag/v2.0_B4__(HPU_3.1_Rev5_B1)
and download the **BOOT.bin** file.
You can find Register information in **HPU Core 3.1.5 - User Guide - RevC.pdf** file.
Please note that the width of bus address has been enlarged to 8 bit (hopefully it doesn't require a software upgrade as well)

**Comment by arrenglover** [comment]
I replaced the BOOT.bin as requested but the driver cannot be insterted:
``IIT-hpu: HPU IP has wrong version: 42303131``

**Comment by mcasti-iit** [comment]
Yes, that is related to Issue 7.
In order to trace the test/debug release and to know what is loaded in hardware, I change that way the ID. It should be useful to make driver bypass the version check if it starts with 42 ( ASCII code for "B", that stands for "Beta")

**Comment by arrenglover** [comment]
so what is the best way to proceed with this test then?

**Comment by francescodiotalevi** [comment]
Ciao guys,
attached the zipped HPU linux driver that bypasses the Version check when it starts with 'B'
**I want to remark that the usage of IOCTL for writing/reading into HPU registers is highly deprecated**. Indeed the driver could be unaware on what the user is performing.
[iit-hpucore-dma.ko.zip](https://github.com/event-driven-robotics/HPU_Core/files/2468956/iit-hpucore-dma.ko.zip)

**Comment by arrenglover** [comment]
> attached the zipped HPU linux driver that bypasses the Version check when it starts with 'B'

Thanks very much.

> **I want to remark that the usage of IOCTL for writing/reading into HPU registers is highly deprecated**.

Can you link me to a repository, where I can get documentation on what the driver defined commands  are and how they are called. I would also like to be able to open issues, as the current command to set the hpu_core in loopback mode wasn't working for me.

**Comment by arrenglover** [comment]
> attached the zipped HPU linux driver that bypasses the Version check when it starts with 'B'

I'm still getting the version mismatch error from dmesg and not getting the device in /dev:

```
[ 3957.276129] IIT-hpu: Probing hpu
[ 3957.276460] IIT-hpu: HPU IP has wrong version: 42303131
```

**Comment by francescodiotalevi** [comment]
Very very strange...
Please Arren try the attached driver. This definitely doesn't check any specific version... It only print the version...
As soon as I have a zedboard system in my hands I'll try again as suggested by @mcasti-iit 
[iit-hpucore-dma.ko.zip](https://github.com/event-driven-robotics/HPU_Core/files/2469058/iit-hpucore-dma.ko.zip)

**Comment by arrenglover** [comment]
> Please Arren try the attached driver. This definitely doesn't check any specific version... It only print the version...

Thanks Francesco, this one opened without any problems

**Comment by francescodiotalevi** [comment]
Arren,
can you check from your terminal if you type dmesg and check if you see something like:
**HPU version is: ??**
Which is the displayed value?

**Comment by arrenglover** [comment]
Same as before:
```
[ 4871.600073] IIT-hpu: Probing hpu
[ 4871.603354] IIT-hpu: HPU version is: 42303131
```

**Comment by francescodiotalevi** [comment]
Thank you Arren.
Maurizio ( @mcasti-iit ) can you check that the hw that Arren is using is really the one that you want to investigate?

**Comment by andreamerello** [comment]
@arrenglover: You are right: something strange happens... I had the HPU working in loop mode, but it turned out that it's just because I've SW1 in "up" position on my board.. 

So I just realized that without that the loop does not work. Have to check what's happening..

**Comment by arrenglover** [comment]
@andreamerello - yes this is the same behaviour I found 👍 

How should I go about reporting any issues, if I come across them?
